### ISSUE TITLE
unpack(): Type N: not enough input, need 4, have 0

### ISSUE THREAD

**Issue by kodekraft** [basho/riak-phppb-client#12]
In the `\Basho\Riak\Api\Pb::readMessageLength` method, I sporadically receive this exception:
`Exception 'ErrorException' with message 'unpack(): Type N: not enough input, need 4, have 0'`

It only occurs when I attempt to send a QueryIndex command, where the data has a binary secondary index.

**Comment by christophermancini** [comment]
This may be related to the issue in the main client. https://github.com/basho/riak-php-client/pull/131

**Comment by christophermancini** [comment]
https://github.com/basho/riak-php-client/pull/131 has been merged, please check to see if you still experience issues if you manually use the develop branch

**Comment by kodekraft** [comment]
Thanks for the update.  I've gone ahead and imported the develop branch and deployed my code base. I'll run my performance testing again and see if this issue occurs again. I'll report back with my findings shortly.

**Comment by kodekraft** [comment]
Hey Chris,

I no longer appear to be receiving the exception in the `\Basho\Riak\Api\Pb::readMessageLength` method using the develop branch of the riak-php-client.

But, I'm now seeing a new exception:
`Exception 'Basho\Riak\Api\Exception' with message 'expected to read 4 bytes, stream_get_contents() returned FALSE'`

being raised in the `\Basho\Riak\Api\Pb::readBytes` method.

Also, received this exception as well:

`Exception 'ErrorException' with message 'fwrite(): send of 63 bytes failed with errno=32 Broken pipe' in basho/riak-pb/src/Riak/Api/Pb.php:318`

**Comment by lukebakken** [comment]
These could be caused by network errors. Are you using a proxy of any kind (like HAProxy)? Is this a development environment? If possible, can you archive the `log/` directory from each of your Riak nodes and attach the archive here?

**Comment by kodekraft** [comment]
Hey Luke, not using any proxying between the app and riak nodes. I believe the last exception that I noted is a result of our network admins doing changes, so i'm going to disregard that one at the moment.  The other issue appears that the connection established is closing for some reason. 

I've overrode the `\Basho\Riak\Api\Pb::debug` method so I can see more of the (breakdown of) communication. 

I'll let you know what I find ...

**Comment by lukebakken** [comment]
Thanks. If you have anything else to add, please re-open this issue.

**Comment by kodekraft** [comment]
Hey Luke, in my previous comment I wasn't completely forthcoming when you asked about using a proxy. I am using php-fpm which does act as a proxy between apache and php.  Interestingly enough, when these errors occur, a restart of the php-fpm service resolves the issue temporarily.

Is this possibly an issue with the PHP protobuf extension that I'm using?
It's the extension from allegro/protobuf package, php7 branch, v0.11.0.

**Comment by lukebakken** [comment]
We don't test `riak-phppb-client` with that protobuf extension, but I don't think that is the issue.

Is this the exception that is resolved by a `php-fpm` restart?

```
Exception 'Basho\Riak\Api\Exception' with message 'expected to read 4 bytes, stream_get_contents() returned FALSE'
```

**Comment by kodekraft** [comment]
Yes.

I think using the PHP `stream_get_meta_data` function should be incorporated into the `Pb::readBytes()` method.  That function should give more information about the state of the socket to help provide more meaningful errors when `stream_get_contents` returns FALSE.

What do you think?

**Comment by kodekraft** [comment]
Using the `stream_get_meta_data` method, I captured more information when that exception is raised:
`
array (
  'timed_out' => true,
  'blocked' => true,
  'eof' => false,
  'stream_type' => 'tcp_socket/ssl',
  'mode' => 'r+',
  'unread_bytes' => 0,
  'seekable' => false,
)
`

**Comment by lukebakken** [comment]
That is interesting information, thanks. It suggests the PB client's timeouts should be adjustable, since reading data from `php-fpm` is timing out.

@heylarson - if I create a branch that implements timeouts, is it easy for you to test in your environment?

http://php.net/manual/en/function.stream-set-timeout.php

**Comment by kodekraft** [comment]
@lukebakken should be no problem.

**Comment by lukebakken** [comment]
@heylarson - please give this branch a try:

https://github.com/basho/riak-phppb-client/tree/fixes/lrb/add-socket-timeouts-gh-12

It sets the read/write timeout for the socket to 5 seconds. I can't seem to find what the default value is for PHP, but I suspect it is system-dependent.

If you still get `'timed_out' => true`, you can increase the stream timeout via this builder method:

https://github.com/basho/riak-php-client/blob/fixes/lrb/add-socket-timeouts/src/Riak/Node/Builder.php#L210-L217

Or, use `setStreamTimeout` on the `Config` object if you are using that.

Please let me know how this works. Thanks!

**Comment by kodekraft** [comment]
Hey Luke, I used your branch in our environment for over a day, and it looks the error still occurs:

`"expected to read 4 bytes, stream_get_contents() returned FALSE. Info: {"timed_out":true,"blocked":true,"eof":false,"stream_type":"tcp_socket\/ssl","mode":"r+","unread_bytes":0,"seekable":false}"`

I didn't change the stream timeout from the default 1 minute since that should be more than enough time to process the request. When the exception occurs, is it possible that the connection handle has been lost?  Should the fix attempt to re-establish the connection?

**Comment by lukebakken** [comment]
I would expect the error to happen less frequently with a 60 second timeout ... is that the case?

`timed_out` is pretty specific - the TCP socket is still open but no data is being sent on it. This means that `php-fpm` or whomever is sending data to the Riak PHP PB client has stopped sending data.

Can you describe what is happening in your application when this error happens? Are there specific Riak operations that result in these timeouts?

**Comment by kodekraft** [comment]
The app is attempting to either fetch objects or query an index every time this exception occurs.

Should I adjust the stream timeout property to something more than 60 seconds, like:
`->withStreamTimeout(60 * 15)`
15 minutes?

Do I need to make any other config changes when setting up the riak client?

Also, I believe you forgot to `return $this` with the `Builder::withStreamConnection()` and `Builder::withConnectionTimeout()` methods.

**Comment by lukebakken** [comment]
Can you log the operation that results in the timeout?

My guess is that it is the index query that is timing out because the client is waiting on data from Riak. If so, is it a secondary index query or Solr?

**Comment by kodekraft** [comment]
I will start logging the operations and post them when the exception happens again.

It's a secondary index.

**Comment by lukebakken** [comment]
Great, I appreciate it.

**Comment by kodekraft** [comment]
Not sure how helpful this is, but this was from enabling `Pb::debug()` debugging.

```
[2016-11-01 08:56:49] local.DEBUG: Basho\Riak\Api\Pb\Message\RpbIndexReq
bucket: "dev.competition"
index: "category_bin"
qtype: 0
key: "sport"
type: "default"

[2016-11-01 08:57:17] local.DEBUG: Basho\Riak\Api\Pb\Message\RpbIndexReq
bucket: "dev.competition"
index: "category_bin"
qtype: 0
key: "sport"
type: "default"

[2016-11-01 08:57:19] local.ERROR: Exception 'Basho\Riak\Api\Exception' with message 'expected to read 4 bytes, stream_get_contents() returned FALSE. Info: {"timed_out":true,"blocked":true,"eof":false,"stream_type":"tcp_socket\/ssl","mode":"r+","unread_bytes":0,"seekable":false}' in /var/www/site/releases/20161031230226/vendor/basho/riak-pb/src/Riak/Api/Pb.php:968.
```

**Comment by lukebakken** [comment]
Thanks for the info. I suspect your best option at this point is to increase the timeout past 60 seconds. There is no way for the client to recover from this timeout since the operation can't be canceled and re-trying would just end up in another timeout (most likely).

If you execute the same secondary index timeout using the HTTP interface, does it also execute slowly?

http://docs.basho.com/riak/kv/2.1.4/developing/usage/secondary-indexes/#querying

You should check the `/var/log/riak` directory on all of your nodes to see if there is anything helpful in the files. If you'd like to archive the logs and attach them here, that's fine.

**Comment by kodekraft** [comment]
I know this issue is now closed, but found an interesting workaround.  Appears that when we configured our load balancer to send heartbeat messages every 5 seconds, we no longer experienced this issue. The heartbeat messages we configured hit a heartbeat endpoint on the application.  This appears to keep php-fpm socket open, thus preventing the timeout issues with the riak pb socket.

I'm now wondering if adjusting the KeepAlive settings for httpd could possibly resolve this issue without the need for the heartbeat solution.
